core/src/Cabana_Experimental_NeighborList.hpp

```diff
@@ -165,17 +164,39 @@ struct NeighborDiscriminatorCallback2D_FirstPass
     }
 };
 
+// Preallocate and attempt fill in the first pass
+template <typename Counts, typename Neighbors, typename Tag>
+struct NeighborDiscriminatorCallback2D_FirstPass_BufferOptimization
+{
+    Counts counts;
+    Neighbors neighbors;
+    using tag = ArborX::Details::InlineCallbackTag;
+    template <typename Predicate>
+    KOKKOS_FUNCTION void operator()( Predicate const &predicate,
+                                     int primitive_index ) const
+    {
+        int const predicate_index = getData( predicate );
+        if ( CollisionFilter<Tag>::keep( predicate_index, primitive_index ) )
+        {
+            if ( counts( predicate_index ) < (int)neighbors.extent( 1 ) )
+            {
+                neighbors( predicate_index, counts( predicate_index )++ ) =
```

Comment:
is this operation on `counts` atomic? does it need to be?

Reply:
See comment
https://github.com/ECP-copa/Cabana/blob/9cf488b7e8ec7dbda6616787e44c9b58390b6054/core/src/Cabana_Experimental_NeighborList.hpp#L211-L212